blob/service.go

```diff
@@ -0,0 +1,302 @@
+	// but we have to guarantee that all our stored subtree roots will be on the same height(e.g. one
+	// level above shares).
+	// TODO(@vgonkivs): rework the implementation to perform all verification without network requests.
+	_, _, err := s.getByCommitment(ctx, height, nID, com)
+	if err != nil {
+		return false, err
+	}
+
+	return true, nil
+}
+
+// getByCommitment retrieves the DAH row by row, fetching shares and constructing blobs in order to
+// compare Commitments. Retrieving is stopped once the requested blob/proof is found.
+func (s *Service) getByCommitment(
+	ctx context.Context,
+	height uint64,
+	nID namespace.ID,
+	commitment Commitment,
+) (*Blob, *Proof, error) {
+	log.Infow("requesting blob",
+		"height", height,
+		"nID", nID.String())
+
+	header, err := s.getByHeight(ctx, height)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var (
+		rawShares = make([]share.Share, 0)
+		proofs    = make(Proof, 0)
+		amount    int
+		blobShare *shares.Share
+	)
+
+	namespacedShares, err := s.shareGetter.GetSharesByNamespace(ctx, header.DAH, nID)
+	if err != nil {
+		return nil, nil, err
+	}
+	for _, row := range namespacedShares {
+		if len(row.Shares) == 0 {
+			break
+		}
+
+		rawShares = append(rawShares, row.Shares...)
+		proofs = append(proofs, row.Proof)
+
+		// reconstruct the `blobShare` from the first rawShare in range
+		// in order to get blob's length(first share will contain this info)
+		if blobShare == nil {
+			for i, sh := range rawShares {
+				bShare, err := shares.NewShare(sh)
+				if err != nil {
+					return nil, nil, err
+				}
+
+				// ensure that the first share is not a NamespacePaddingShare
+				// these shares are used to satisfy the non-interactive default rules
+				// and are not the part of the blob, so should be removed.
+				isPadding, err := bShare.IsPadding()
+				if err != nil {
+					return nil, nil, err
+				}
+				if isPadding {
+					continue
+				}
+
+				blobShare = &bShare
+				// save the length.
+				length, err := blobShare.SequenceLen()
+				if err != nil {
+					return nil, nil, err
+				}
+				amount = shares.SparseSharesNeeded(length)
+				rawShares = rawShares[i:]
+				break
+			}
+		}
+
+		// move to the next row if the blob is incomplete.
+		if amount > len(rawShares) {
+			continue
+		}
+
+		// reconstruct the Blob.
+		blob, err := sharesToBlobs(rawShares[:amount])
+		if err != nil {
+			return nil, nil, err
+		}
+
+		// compare commitments.
+		if bytes.Equal(blob[0].Commitment(), commitment) {
+			return blob[0], &proofs, nil
+		}
+
+		// drop info of the checked blob
+		rawShares = rawShares[amount:]
+		blobShare = nil
+	}
+	return nil, nil, ErrBlobNotFound
+}
+
+// getBlobs retrieves the DAH and fetches all shares from the requested namespace.ID and converts
+// them to Blobs.
+func (s *Service) getBlobs(ctx context.Context, nID namespace.ID, root *share.Root) ([]*Blob, error) {
+	namespacedShares, err := s.shareGetter.GetSharesByNamespace(ctx, root, nID)
```

Comment:
We should be careful with ErrNotFound errors here. And map them to blob not found. 
cc @distractedm1nd and @walldiss

Reply:
Applies to all GetSharesByNamespace calls

Reply:
The current way is fine - we should just be aware of this when changing the returned error for ErrNamespaceNotFound to be empty data instead of an error.